Packet++/header/LdapLayer.h

```diff
@@ -0,0 +1,267 @@
+		 * It accepts the getter method and an out variable. It tries to call the getter and if no exception
+		 * is thrown, the out variable will contain the result.
+		 *
+		 * Here is an example:
+		 * @code
+		 * uint16_t messageId;
+		 * ldapLayer->tryGet(&pcpp::LdapLayer::getMessageID, messageId));
+		 * @endcode
+		 *
+		 * We call getMessageID(), if no exception is thrown the variable messageId will hold the result
+		 *
+		 * @tparam Method The class method type
+		 * @tparam ResultType The expected result type (for example: uint8_t, std::string, etc.)
+		 * @param[in] method The class method to call
+		 * @param[out] result An outvariable to contain the result if no exception is thrown
+		 * @return True if no exception was thrown or false otherwise
+		 */
+		template <typename Method, typename ResultType>
+		bool tryGet(Method method, ResultType& result)
+		{
+			return internalTryGet(this, method, result);
+		}
+
+		/**
+		 * A static method that checks whether a source or dest port match those associated with the LDAP protocol
+		 * @param[in] port The port number to check
+		 * @return True if this is an LDAP port, false otherwise
+		 */
+		static bool isLdapPort(uint16_t port) { return port == 389; }
+
+		/**
+		 * A static message to parse an LDAP message from raw data
+		 * @param[in] data A pointer to the raw data
+		 * @param[in] dataLen Size of the data in bytes
+		 * @param[in] prevLayer A pointer to the previous layer
+		 * @param[in] packet A pointer to the Packet instance where layer will be stored in
+		 * @return An instance of LdapLayer if this is indeed an LDAP message, nullptr otherwise
+		 */
+		static LdapLayer* parseLdapMessage(uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+
+		// implement abstract methods
+
+		/**
+		 * Tries to identify more LDAP messages in this packet if exist
+		 */
+		void parseNextLayer() override;
+
+		/**
+		 * @return The size of the LDAP message
+		 */
+		size_t getHeaderLen() const override { return m_Asn1Record->getTotalLength(); }
+
+		void computeCalculateFields() override {}
+
+		OsiModelLayer getOsiModelLayer() const override { return OsiModelApplicationLayer; }
+
+		std::string toString() const override;
+
+	protected:
+		std::unique_ptr<Asn1Record> m_Asn1Record;
+
+		LdapLayer(std::unique_ptr<Asn1Record>& asn1Record, uint8_t* data, size_t dataLen, Layer* prevLayer, Packet* packet);
+		LdapLayer() = default;
+		void init(uint16_t messageId, LdapOperationType operationType, const std::vector<Asn1Record*>& messageRecords, const std::vector<LdapControl>& controls);
+		virtual std::string getExtendedStringInfo() const { return ""; }
+
+		template <typename LdapClass, typename Method, typename ResultType>
+		bool internalTryGet(LdapClass* thisPtr, Method method, ResultType& result)
+		{
+			try
+			{
+				result = (thisPtr->*method)();
+				return true;
+			}
+			catch (...)
+			{
+				return false;
+			}
+		}
```

Comment:
This function can be made `static`. Also why can't this use the implicit `this` pointer?

Reply:
I'd like to implement `tryGet` for `LdapLayer` and future layer classes that will derive from `LdapLayer`. I tried different options and this is the only way I could make it work. Can you suggest a better way?

You can find my draft branch with implementation of more LDAP message types here: https://github.com/seladb/PcapPlusPlus/tree/ldap

Reply:
Can't really think of any other ways off the top of my head, so I guess it is fine.

One suggestion for better readability to replace 
`result = (thisPtr->*method)();` with `result = std::mem_fn(method)(thisPtr);`,
as I was mildly confused what was happening on that line when I read it the first time.
Requires `<functional>`.

Reply:
sure, added in this commit: https://github.com/seladb/PcapPlusPlus/pull/1412/commits/4c075713ca22909d62b0cd54a88a239016d49139